<a href="https://colab.research.google.com/github/paragpujari/ASSIGNMENT-1-Basic-Python-Concepts/blob/main/9_Sentence_Embeddings_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### **SENTENCE EMBEDDINGS:**

Sentence Embedding refers to the technique of converting the sentence text into the dense numerical vector that captures the semantic meaning.

It makes use of the context and semantics of the full sentence.


#### **Steps used in this Algorithm:--**

1.  Import all the necessary libraries

2.  Load the Pretrained Model

3.  Create the Sample Sentences

4.  Generate the  Sentence Embeddings

5.  Check the Embedding Shape

6.  View One Embedding

7.  Compute Similarity Between Sentences

In [130]:
!pip install nbstripout
! nbstripout your_notebook.ipynb

Could not strip 'your_notebook.ipynb': file not found


In [131]:
import nbformat

# Load notebook
nb = nbformat.read("9-Sentence_Embedding_NLP.ipynb", as_version=4)

# Remove widgets metadata if present
if "widgets" in nb["metadata"]:
    del nb["metadata"]["widgets"]

# Save cleaned notebook
nbformat.write(nb, "/content/9-Sentence_Embedding_NLP.ipynb")

print("Clean notebook saved successfully!")

Clean notebook saved successfully!


### Step 1: Import all the necessary libraries

In [132]:
import  numpy            as   np
import  pandas           as   pd
import  matplotlib.pyplot as   plt
import  seaborn           as   sns

from    sentence_transformers      import SentenceTransformer
import  nltk

from    nltk.tokenize               import  sent_tokenize, word_tokenize
from    sklearn.metrics.pairwise   import cosine_similarity

### **OBSERVATIONS:**

1.  numpy -------------->  Computation of the numerical array

2.  pandas ------------->  Data Creation and Manipulation

3.  matplotlib --------->  Data Visualization

4.  seaborn ------------>  Data Correlation

5.  SentenceTransformer ------> loads a pre trained transformer model required for sentence embeddings

6.  sent_tokenize ----------->  converts the paragraphs into sentences

7.  word_toeknize ---------->  converts the sentences into words

8.  cosine_similarity ------->  get the similarity ratio between the sentences

### Step 2: Load the Pretrained Model

In [133]:
### Load the Pre trained BERT Model

model = SentenceTransformer('All-MiniLM-L6-v2')

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/All-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


### **OBSERVATIONS:**

1.  **All-MiniLM-L6-v2** is a light weighted pre trained BERT Model.

2.  It produces 384 dimensional embeddings.

3.  It is fast and accurate for most of the tasks.


### **Why do we use the model here?**

1. The sentence embeddings requires BERT like pre trained transformer model.

2. It needs huge number of parameter for training.

3. It can be trained on massive datasets.

4. It requires huge computational power.

5. It produces high dimensional embeddings.

### Step 3: Create the Sample Sentences

In [134]:
sentences = [
    "I love machine learning",
    "I enjoy studying artificial intelligence",
    "The weather is very hot today",
    "Deep learning is a branch of AI"
]

In [135]:
print(sentences)

['I love machine learning', 'I enjoy studying artificial intelligence', 'The weather is very hot today', 'Deep learning is a branch of AI']


### **OBSERVATIONS:**

1. A corpus is created that has four sentences.

### Step 4: Generate the  Sentence Embeddings

In [136]:
### To generate the sentence embeddings

embeddings = model.encode(sentences, normalize_embeddings=True)

In [137]:
embeddings

array([[-0.04363636, -0.05905438,  0.08201232, ...,  0.09771022,
         0.0168808 , -0.09040585],
       [-0.00688083, -0.05725535,  0.02433584, ...,  0.07118098,
         0.02248301, -0.06988683],
       [-0.0017241 ,  0.07078548,  0.0869412 , ..., -0.00364969,
        -0.14153752,  0.07740253],
       [-0.04006512, -0.11033705,  0.03086832, ...,  0.0314955 ,
         0.03731575, -0.05878681]], dtype=float32)

###**OBSERVATIONS:**

1.  The sentence corpus is taken as an input.

2.  It is first tokenized

3.  Then it is passed through the transformer model.

4.  Max Pooling is performed

5. Then the fixed sized dense dimensional vectors is formed.

6. There were four sentences. After applying sentence embeddings on it,

    (a.)  It is converted into four rows. Each row has the dense dimensional vectors.

    (b.) There are a total of 384 dense dimensional vectors in each row.

### Step 5: Check the Embedding Shape

In [138]:
print("Shape of the embeddings is:", embeddings.shape)

Shape of the embeddings is: (4, 384)


### **OBSERVATIONS:**

1. The shape of the sentence embeddings is (4, 384)

    (a.)   4   ---------->   sentences

    (b.)   384 ---------->   features

    So, the shape of the embeddings is (total sentences, total features)

### Step 6: View One Embedding

In [139]:
print("First Embedding is:", embeddings[0])

First Embedding is: [-4.36363593e-02 -5.90543821e-02  8.20123181e-02 -1.07671609e-02
  6.11958653e-02 -5.51596954e-02 -2.90437229e-03 -3.67232487e-02
  2.13971715e-02 -7.56041985e-03 -7.17570186e-02  5.08238301e-02
  3.80331874e-02 -1.62503012e-02 -3.21699679e-02  1.52751934e-02
 -4.67937700e-02  1.71293058e-02 -1.20174505e-01 -1.22229166e-01
 -8.08737576e-02  1.70528069e-02 -2.08724458e-02  5.68292104e-03
  1.60145143e-03  2.62672398e-02  2.99523007e-02 -2.75703310e-03
  7.17105158e-03 -6.12753667e-02 -3.61441029e-03  2.48506572e-02
  2.56387256e-02  4.12390120e-02 -1.02098763e-01  8.36896058e-03
  1.39943417e-02  3.09881680e-02  2.43119933e-02  3.08077447e-02
 -4.40565869e-02 -1.38978288e-02  1.30948192e-02  3.64952255e-03
  7.49757737e-02  8.54910016e-02 -5.81432953e-02 -7.58069381e-02
  3.96832153e-02  5.06207310e-02 -8.65324214e-02 -3.23279314e-02
 -2.03664098e-02 -1.62433330e-02 -5.87623119e-02  2.10438650e-02
  6.54709041e-02 -1.25287939e-02  3.35284434e-02 -2.76364014e-02
  2.9

###**OBSERVATIONS:**

1. It corresponds to the first sentence having 384 features.

2. It even depicts about the semantic meaning in the vector space.

### Step 7: Compute Similarity Between Sentences

In [140]:
from sklearn.metrics.pairwise  import cosine_similarity

### get the similarity ratio between the sentences

similarity_matrix=cosine_similarity(embeddings)

In [141]:
similarity_matrix

array([[ 1.0000001 ,  0.56718457,  0.02401708,  0.5580992 ],
       [ 0.56718457,  0.99999994, -0.04725922,  0.47398493],
       [ 0.02401708, -0.04725922,  1.0000001 , -0.06026478],
       [ 0.5580992 ,  0.47398493, -0.06026478,  1.0000001 ]],
      dtype=float32)

### **OBSERVATIONS:**

1.  It depicts about the similarity ratio between the sentences in the corpus.

     Similarity ratio between sentence 1 and 1  ------>  1 (Totally same)
   
     Similarity ratio between sentence 1 and 2 -------> 0.56 (half same)
